src/storage/src/storage_state.rs

```diff
@@ -965,16 +965,51 @@ impl<'w, A: Allocate> Worker<'w, A> {
                 }
                 StorageCommand::RunIngestions(ingestions) => {
                     ingestions.retain_mut(|ingestion| {
+                        // Subsources can be dropped independently of their
+                        // primary source, so we evaluate them in a separate
+                        // loop.
+                        for export_id in ingestion
+                            .description
+                            .source_exports
+                            .keys()
+                            .filter(|export_id| **export_id != ingestion.id)
+                        {
+                            if drop_commands.remove(export_id)
+                                || self.storage_state.dropped_ids.contains(&ingestion.id)
+                            {
+                                self.storage_state.dropped_ids.insert(*export_id);
+                            }
+                        }
+
                         if drop_commands.remove(&ingestion.id)
                             || self.storage_state.dropped_ids.contains(&ingestion.id)
                         {
-                            // Make sure that we report back that the ID was
-                            // dropped.
-                            self.storage_state.dropped_ids.insert(ingestion.id);
+                            // If an ingestion is dropped, so too must all of
+                            // its subsources (i.e. ingestion exports, as well
+                            // as its progress subsource).
+                            for id in ingestion.description.subsource_ids() {
```

Comment:
small nit that's out of scope for this PR but I wanted to highlight:

The naming of the `subsource_ids()` helper seems odd that it includes the remap-collection.

Whereas the line above `source_exports.keys().filter(|export_id| **export_id == !ingestion.id)` fits what I'd expect `subsource_ids()` to return.

Maybe it's because I have a different mental model of how you perceive these terms. It's tricky since we use different names for the concept of 'the collections downstream of a given source': `SourceExport`, `IngestionExport`, `subsource`. (and sometimes we refer to them more generally as the persist-shards and collections of a source, which is factual but adds to the mental overhead).

Reply:
If you type `SHOW SUBSOURCES ON foo`, it includes the remap collection. However, it doesn't show the primary collection. All of this is in flux and a bit of a mess; hopefully deprecating the concept of subsources entirely will resolve things a bit.